Sprint 14

Dark2801 opened the pull request:
Полностью переписал проект. Задание 14 спринта

Nikitaaeee commented on Tracker/Cells/ListCell.swift:
```diff
@@ -0,0 +1,99 @@
+        return button
+    }()
+    
+    // MARK: - Properties
+    static let identifier = "ListCell"
+    
+    // MARK: - Lifecycle
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+        
+        configureViews()
+        configureConstraints()
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    // MARK: - Methods
+    func configure(label: String, value: String?, position: ListOfItems.Position) {
```
Круто, что ты используешь комментарии `// MARK: - ...` ! 🔥

Dark2801 replied:
Спасибо!

Nikitaaeee commented on Tracker/Cells/TrackerCell.swift:
```diff
@@ -0,0 +1,164 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func prepareForReuse() {
+        super.prepareForReuse()
+        tracker = nil
+        days = 0
+        addDayButton.setImage(UIImage(systemName: "plus"), for: .normal)
+        addDayButton.layer.opacity = 1
+    }
```
👍🏻👍🏻👍🏻

Nikitaaeee commented on Tracker/Extensions/Colors+Extensions.swift:
```diff
@@ -0,0 +1,55 @@
+        UIColor(named: "ColorSelection17")!,
+        UIColor(named: "ColorSelection18")!
+    ]
+}
+
+    
+    
+    
+    
+    
+    
+    
+    
+    
+
```
Пустые строки в конце классов/структур/методов можно удалять
**SwiftLint** (сервис для чистого кода), например, может ругаться на такие места (_кидать warning_)

Dark2801 replied:
Удалил!

Nikitaaeee commented on Tracker/Extensions/Date Int+Extension.swift:
```diff
@@ -0,0 +1,37 @@
+//
+//  Date Int+Extension.swift
+//  Tracker
+//
+//  Created by Андрей Мерзликин on 05.11.2023.
+//
+
+import Foundation
+
+extension Date {
+    static func from(year: Int, month: Int, day: Int) -> Date? {
+        let calendar = Calendar(identifier: .gregorian)
+        var dateComponents = DateComponents()
+        dateComponents.year = year
+        dateComponents.month = month
+        dateComponents.day = day
+        return calendar.date(from: dateComponents) ?? nil
+    }
+    
+    static func from(date: Date) -> Date? {
+        let components = Calendar.current.dateComponents([.year, .month, .day], from: date)
+        guard let day = components.day, let month = components.month,
+              let year = components.year else { return nil }
+        return from(year: year, month: month, day: day)
+    }
+}
+
+extension Int {
+    func days() -> String {
+        var ending: String!
+        if "1".contains("\(self % 10)")       { ending = "день"}
+        if "234".contains("\(self % 10)")     { ending = "дня" }
+        if "567890".contains("\(self % 10)")  { ending = "дней"}
+        if 11...14 ~= self % 100              { ending = "дней"}
+        return "\(self)" + ending
+    }
+}
```
**extension**'ы на **Int** и **Date** можно разнести по разным файлам, обычно каждому объекту создают свой файл для расширений.

Nikitaaeee commented on Tracker/Helpers/Emoji.swift:
```diff
@@ -0,0 +1,15 @@
+//  Created by Андрей Мерзликин on 05.11.2023.
+//
+
+import Foundation
+
+let emojisArray = [
+    "🙂", "😻", "🌺", "🐶", "❤️", "😱",
+    "😇", "😡", "🥶", "🤔", "🙌", "🍔",
+    "🥦", "🏓", "🥇", "🎸", "🏝", "😪",
+]
+
```
Заметил, что ты хранишь список эмоджи в **глобальной переменной**. 
Это может быть удобно, но не всегда хорошо для организации кода. Так легко случайно что-то поменять где не надо и сложнее управлять памятью. 

Лучше обернуть эмоджи в какой-либо объект и использовать там, где они нужны, так твой код будет более чистым и понятным. 
А также так легче будет с тестами и поддержкой кода в будущем! 💡

Nikitaaeee commented on Tracker/Trackers/TrackerCategoryNames.swift:
```diff
@@ -0,0 +1,41 @@
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        
+        addSubview(titleLabel)
+        titleLabel.translatesAutoresizingMaskIntoConstraints = false
+        
+        NSLayoutConstraint.activate([
+            titleLabel.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 24),
+            titleLabel.topAnchor.constraint(equalTo: topAnchor),
+            titleLabel.bottomAnchor.constraint(equalTo: bottomAnchor),
+        ])
```
В других местах в коде у тебя используются два отдельных метода для _добавления view на супервью_ и _настройки констрейнтов_ ( `configureViews()` и `configureConstraints()` ), было бы здорово вынести этот код из инициализатора в такие же методы.

Повторяющийся и единообразный кодстайл помогает коду быть более читаемым и предсказуемым, что очень важно при работе в команде 😊

Nikitaaeee commented on Tracker/UIStructs/ListOfItems.swift:
```diff
@@ -0,0 +1,75 @@
+        translatesAutoresizingMaskIntoConstraints = false
+        layer.cornerRadius = 16
+        backgroundColor = .BackgroundDay
+        
+        addSubview(border)
+        border.translatesAutoresizingMaskIntoConstraints = false
+        
+        NSLayoutConstraint.activate([
+            border.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 16),
+            border.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -16),
+            border.bottomAnchor.constraint(equalTo: bottomAnchor),
+            border.heightAnchor.constraint(equalToConstant: 0.5)
+        ])
+    }
```
И тут тоже

Nikitaaeee commented on Tracker/UIStructs/WeekDay.swift:
```diff
@@ -0,0 +1,40 @@
+            let second = Self.allCases.firstIndex(of: rhs)
+        else { return false }
+        
+        return first < second
+        
+    }
```
Пустую строку после **return** можно убрать

Dark2801 replied:
Убрал

Nikitaaeee commented on Tracker/ViewControllers/StatisticViewController.swift:
```diff
@@ -0,0 +1,72 @@
+        stack.spacing = 8
+        return stack
+    }()
+    
+    // MARK: - Lifecycle
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        configureViews()
+        configureConstraints()
+        
+    }
+}
+//MARK: - Extensions, Layout methods
+private extension StatisticViewController {
```
В некоторых местах в работе у тебя есть пустая строка **перед** комментариями `// MARK: - ...`, где-то есть пустая строка **после**, а где-то пустых строк **нет ни сверху ни снизу**.

Было бы здорово выбрать один из вариантов проставления отступов/пустых строк и следовать ему при написании кода. (Например,  оставлять пустую строку над и под комментарием `// MARK: - ...`)

Dark2801 replied:
Исправил!

Dark2801 committed:
Добавил Colors and Icons

Dark2801 committed:
Создал LaunchScreen

Dark2801 committed:
Создал структуру Helpers и добавил файл UIImage+Extension
в нем прописал кодом все UIImage

Dark2801 committed:
Прописал шрифты в коде UIFont+Extensions

Dark2801 committed:
Прописал UIColor в код. Создал UIColor+Extensions

Dark2801 committed:
Создал группу TrackesTab

Dark2801 committed:
TrackersTab -> TrackersViewController

Dark2801 committed:
Created TrackersViewController

Dark2801 committed:
Created TrackerSearchBar

Dark2801 committed:
New Group TrackersCollectionView

Dark2801 committed:
Created TrackerCollectionViewCell

Dark2801 committed:
Created TrackerCollectionViewSectionHeader

Dark2801 committed:
Created TrackerCollectionViewDataSource

Dark2801 committed:
Created TrackerCollectionViewDelegate

Dark2801 committed:
Created GeometricParams

Dark2801 committed:
Created new group Models -> Tracker

Dark2801 committed:
Created Models -> WeekDay

Dark2801 committed:
Created Helpers -> PlaceholderView

Dark2801 committed:
Created -> Services

Dark2801 committed:
Created  DataManager

Dark2801 committed:
Created Models -> TrackerRecord

Dark2801 committed:
Created -> TrackerCategory

Dark2801 committed:
CreateTrackerViewController

Dark2801 committed:
Создал UIButton+Extensions
Создал RegularTrackerViewController
Изменил AppDelegate и SceneDelegate
Cоздал TabBarController

Dark2801 committed:
Создал RegularTrackerTableViewSubtitleCell

Dark2801 committed:
Создал RegularTrackerTableViewDataSource

Dark2801 committed:
Создал RegularTrackerTableViewDelegate

Dark2801 committed:
Создал новую группу IrregularTrackerViewController

Dark2801 committed:
Создан IrregularTrackerViewController

Dark2801 committed:
Создана группа IrregularTrackerTableView
Создан IrregularTrackerTableViewSubtitleCell
Cоздан IrregularTrackerTableViewDataSource
Создан IrregularTrackerTableViewDelegate

Dark2801 committed:
Созданы СategoryViewController
Создан CategoryTableViewCell
Создан CategoryTableViewDataSource
Cоздан CategoryTableViewDelegate

Dark2801 committed:
Прописаны шрифты UITextField

Dark2801 committed:
Cозданы CreateCategoryViewController
ScheduleViewController
ScheduleTabViewCell
ScheduleTabViewDataSource

Dark2801 committed:
Создан StatisticsViewController

Dark2801 committed:
Добавлены Emoji

Dark2801 committed:
Добавил Emoji

Dark2801 committed:
Добавил иконку

Dark2801 committed:
Выровнил код

Dark2801 committed:
Удалено лишнее

Nikitaaeee commented on Tracker/Services/DataManager.swift:
```diff
@@ -0,0 +1,60 @@
+    static let shared = DataManager()
+
+    
+    var categories: [TrackerCategory] = [
```
Можно убрать вторую пустую строку, одной достаточно :)

Dark2801 replied:
Исправил

Nikitaaeee commented on Tracker/TrackersTab/IrregularTrackerViewController/IrregularTrackerViewController.swift:
```diff
@@ -0,0 +1,287 @@
+        ]
+    }
+    
+    func addStackView() {
+        view.addSubview(stackView)
+        stackView.addArrangedSubview(trackerTitleTextField)
+        stackView.addArrangedSubview(symbolsConstraintLabel)
+        
+        NSLayoutConstraint.activate([
+            stackView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 24),
+            stackView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
+            stackView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -16)
+        ])
+    }
+    
+    func addTableView() {
+        view.addSubview(tableView)
+        
+        NSLayoutConstraint.activate([
+            tableView.heightAnchor.constraint(equalToConstant: 75),
+            tableView.topAnchor.constraint(equalTo: stackView.bottomAnchor, constant: 24),
+            tableView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
+            tableView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -16)
+        ])
+    }
+    
+    func addButtonsStackView() {
+        view.addSubview(buttonsStackView)
+        buttonsStackView.addArrangedSubview(cancelButton)
+        buttonsStackView.addArrangedSubview(createButton)
+        
+        NSLayoutConstraint.activate([
+            buttonsStackView.bottomAnchor.constraint(equalTo: view.bottomAnchor, constant: -34),
+            buttonsStackView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 20),
+            buttonsStackView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -20)
+        ])
```
Чаще принято организовывать код настройки UI элементов таким образом:

- Создавать приватный метод _addSubviews()_, в котором добавить все нужные элементы:
`view.addSubview(label)`
- Создавать приватный метод _setupConstraints()_, в котором добавить все нужные констрейнты: 
```
NSLayoutConstraint.activate([
   label.centerYAnchor.constraint(equalTo: view.centerYAnchor)
   ...
])
```
- Вызвать эти два метода во _viewDidLoad_() 

Это совсем не критическое замечание, просто на заметку 😊

Nikitaaeee commented on Tracker/TrackersTab/TrackersViewController/TrackersViewController.swift:
```diff
@@ -0,0 +1,357 @@
+        }
+    
+    
+    private var visibleCategories: [TrackerCategory] = []
```
Тут можно чуть поправить отступ (Выделить код и нажать `Control + I`), а также удалить лишнюю пустую строку.

Dark2801 replied:
Спасибо! Исправил!

Nikitaaeee reviewed (approved):
👍🏻👍🏻👍🏻

Dark2801 committed:
Исправления по ревью
